pytests/test_types.py

```diff
@@ -0,0 +1,458 @@
+    diff = DeepDiff(converted_trade, TRADE, ignore_order=True)
+    assert not diff
+
+
+def test_order_detail_to_pb() -> None:
+    """Test the client order detail type conversions to protobuf."""
+    converted_order_detail = ORDER_DETAIL.to_pb()
+
+    assert isinstance(converted_order_detail, electricity_trading_pb2.OrderDetail)
+    diff = DeepDiff(converted_order_detail.order, ORDER_DETAIL_PB.order)
+    assert not diff
+
+
+def test_order_detail_from_pb() -> None:
+    """Test the client order detail type conversion from protobuf."""
+    converted_order_detail = OrderDetail.from_pb(ORDER_DETAIL_PB)
+
+    assert isinstance(converted_order_detail, OrderDetail)
+    diff = DeepDiff(converted_order_detail.order, ORDER_DETAIL.order)
+    assert not diff
+
+
+def test_gridpool_order_filter_to_pb() -> None:
+    """Test the client gridpool type conversion to protobuf."""
+    converted_gridpool_order_filter = GRIDPOOL_ORDER_FILTER.to_pb()
+
+    assert isinstance(
+        converted_gridpool_order_filter, electricity_trading_pb2.GridpoolOrderFilter
+    )
+    diff = DeepDiff(
+        converted_gridpool_order_filter, GRIDPOOL_ORDER_FILTER_PB, ignore_order=True
+    )
+    assert not diff
+
+
+def test_gridpool_order_filter_from_pb() -> None:
+    """Test the client gridpool type conversion from protobuf."""
+    converted_gridpool_order_filter = GridpoolOrderFilter.from_pb(
+        GRIDPOOL_ORDER_FILTER_PB
+    )
+
+    assert isinstance(converted_gridpool_order_filter, GridpoolOrderFilter)
+    diff = DeepDiff(
+        converted_gridpool_order_filter, GRIDPOOL_ORDER_FILTER, ignore_order=True
+    )
+    assert not diff
+
+
+def test_update_order_to_pb() -> None:
+    """Test the client update order type conversion to protobuf."""
+    converted_update_order = UPDATE_ORDER.to_pb()
+
+    assert isinstance(
+        converted_update_order,
+        electricity_trading_pb2.UpdateGridpoolOrderRequest.UpdateOrder,
+    )
+    diff = DeepDiff(converted_update_order, UPDATE_ORDER_PB, ignore_order=True)
+    assert not diff
+    # Make sure the number of fields in original and converted are the same
+    assert len(converted_update_order.ListFields()) == len(UPDATE_ORDER_PB.ListFields())
+
+
+def test_update_order_from_pb() -> None:
+    """Test the client update order type conversion from protobuf."""
+    converted_update_order = UpdateOrder.from_pb(UPDATE_ORDER_PB)
+
+    assert isinstance(converted_update_order, UpdateOrder)
+    diff = DeepDiff(converted_update_order, UPDATE_ORDER, ignore_order=True)
+    assert not diff
+    # Make sure the number of non-None attributes in original and converted are the same
+    non_none_attrs_converted = sum(
+        1 for v in vars(converted_update_order).values() if v is not None
+    )
+    non_none_attrs_original = sum(
+        1 for v in vars(UPDATE_ORDER).values() if v is not None
+    )
+    assert non_none_attrs_converted == non_none_attrs_original
+
+
```

Comment:
Maybe the same point as in `test_client.py` could be useful here. I would add some tests where we expect the types to not match. This might be a bit tougher for the types than in the client case though so not sure if it makes a lot of sense here.
For example there was a check in place [here](https://github.com/frequenz-floss/frequenz-api-electricity-trading/blob/a66061ec3a702becdafb1f1d18a22d5f20fa075a/py/frequenz/client/electricity_trading/_types.py#L959) if the order has a `valid_until` type which could be tested here also.

Reply:
(fyi I added some tests for no/ invalid timezones in`OrderDetail`)

> For example there was a check in place [here](https://github.com/frequenz-floss/frequenz-api-electricity-trading/blob/a66061ec3a702becdafb1f1d18a22d5f20fa075a/py/frequenz/client/electricity_trading/_types.py#L959) if the order has a valid_until type which could be tested here also.

Here I check that the attributes converted are the right ones (with `deepDiff`), and I make sure that no other instance var is converted (by checking the number of attributes in the converted type matches the one in the original type). You mean I should do multiple of these tests for different fields to update ?

Reply:
> You mean I should do multiple of these tests for different fields to update ?

Yeah ideally but I think we can always increase the test coverage so maybe this is enough for now if we want to focus on deploying soon.